pymongo/change_stream.py

```diff
@@ -16,48 +16,125 @@
 
 import copy
 
+from bson.son import SON
+
+from pymongo import common
+from pymongo.collation import validate_collation_or_none
+from pymongo.command_cursor import CommandCursor
 from pymongo.errors import (ConnectionFailure, CursorNotFound,
                             InvalidOperation, PyMongoError)
 
 
 class ChangeStream(object):
-    """A change stream cursor.
-
-    Should not be called directly by application developers. Use
-    :meth:`~pymongo.collection.Collection.watch` instead.
+    """The change stream cursor abstract base class.
 
-    .. versionadded: 3.6
-    .. mongodoc:: changeStreams
+    Defines the interface for change streams. Should be subclassed to
+    implement the `ChangeStream._create_cursor` abstract method and
+    the `ChangeStream._database` abstract property.
     """
-    def __init__(self, collection, pipeline, full_document,
-                 resume_after=None, max_await_time_ms=None, batch_size=None,
-                 collation=None, session=None):
-        self._collection = collection
+    def __init__(self, target, pipeline, full_document, resume_after,
+                 max_await_time_ms, batch_size, collation,
+                 start_at_operation_time, session):
+        if pipeline is None:
+            pipeline = []
+        elif not isinstance(pipeline, list):
+            raise TypeError("pipeline must be a list")
+
+        common.validate_string_or_none('full_document', full_document)
+        validate_collation_or_none(collation)
+        common.validate_non_negative_integer_or_none("batchSize", batch_size)
+
+        self._target = target
         self._pipeline = copy.deepcopy(pipeline)
         self._full_document = full_document
         self._resume_token = copy.deepcopy(resume_after)
         self._max_await_time_ms = max_await_time_ms
         self._batch_size = batch_size
         self._collation = collation
+        self._start_at_operation_time = start_at_operation_time
         self._session = session
         self._cursor = self._create_cursor()
 
-    def _full_pipeline(self):
+    @property
+    def _aggregation_target(self):
+        """The argument to pass to the aggregate command."""
+        raise NotImplementedError
+
+    @property
+    def _database(self):
+        """The database against which the aggregation commands for
+        this ChangeStream will be run. """
+        raise NotImplementedError
+
+    def _full_pipeline(self, inject_options=None):
```

Comment:
inject_options reads like a verb. Maybe just call it 'options', then in the method do:

options = options or {}

Reply:
Oh, I see why this won't work. Well, I still think you should pick a noun for the argument name. Maybe just 'extra_options'?